File: backend/app/api/v1/routes/assets.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.authz.deps import get_authz_engine, require_permission
from app.authz.engine import AuthzEngine
from app.authz.enums import Action, FeatureKey
from app.core.pagination import Page
from app.database import get_db
from app.models.employee import Employee
from app.schemas.asset import (
    AssetAssignIn,
    AssetAssignmentOut,
    AssetCreate,
    AssetOut,
    AssetReturnIn,
)
from app.services import asset_service

router = APIRouter(tags=["assets"])
FEATURE = FeatureKey.ASSET_MANAGEMENT


@router.get("/assets", response_model=Page)
async def list_assets(
    page: int = Query(1, ge=1),
    page_size: int = Query(25, ge=1, le=200),
    db: AsyncSession = Depends(get_db),
    _employee: Employee = Depends(require_permission(FEATURE, Action.VIEW)),
):
    items, total = await asset_service.list_assets(db, offset=(page - 1) * page_size, limit=page_size)
    return Page(items=items, total=total, page=page, page_size=page_size)


@router.post("/assets", response_model=AssetOut, status_code=201)
async def create_asset(
    payload: AssetCreate,
    db: AsyncSession = Depends(get_db),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.CREATE)),
):
    return await asset_service.create_asset(db, current_employee, payload)


@router.post("/assets/{asset_id}/assign", response_model=AssetAssignmentOut, status_code=201)
async def assign_asset(
    asset_id: uuid.UUID,
    payload: AssetAssignIn,
    db: AsyncSession = Depends(get_db),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.EDIT)),
):
    return await asset_service.assign_asset(db, current_employee, asset_id, payload)


@router.post("/asset-assignments/{assignment_id}/return", response_model=AssetAssignmentOut)
async def return_asset(
    assignment_id: uuid.UUID,
    payload: AssetReturnIn,
    db: AsyncSession = Depends(get_db),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.EDIT)),
):
    return await asset_service.return_asset(db, current_employee, assignment_id, payload)


@router.get("/employees/{employee_id}/assets", response_model=list[AssetAssignmentOut])
async def employee_asset_history(
    employee_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    engine: AuthzEngine = Depends(get_authz_engine),
    current_employee: Employee = Depends(require_permission(FEATURE, Action.VIEW)),
):
    return await asset_service.employee_asset_history(db, engine, current_employee, employee_id)

File: frontend/src/api/assets.ts
import { apiClient } from "./client";
import type { Asset, AssetAssignment, Page } from "./types";

export const assetsApi = {
  list: (page = 1, pageSize = 50) =>
    apiClient.get<Page<Asset>>("/assets", { params: { page, page_size: pageSize } }).then((r) => r.data),
  create: (payload: Partial<Asset>) => apiClient.post<Asset>("/assets", payload).then((r) => r.data),
  assign: (assetId: string, payload: { employee_id: string; assigned_date: string; condition_notes?: string }) =>
    apiClient.post<AssetAssignment>(`/assets/${assetId}/assign`, payload).then((r) => r.data),
  returnAsset: (assignmentId: string, payload: { returned_date: string; condition_notes?: string }) =>
    apiClient
      .post<AssetAssignment>(`/asset-assignments/${assignmentId}/return`, payload)
      .then((r) => r.data),
  employeeHistory: (employeeId: string) =>
    apiClient.get<AssetAssignment[]>(`/employees/${employeeId}/assets`).then((r) => r.data),
};
